Introduces PluginSettingsSidebar slotfill to the Document sidebar.

ryanwelcher opened the pull request:
## Description
Introduce a new slotFill below Status & Availability panel in the Document sidebar.
Closes #13357 

## How has this been tested?
This has been testing locally and there are tests on the PR.

## Screenshots
![slotfill](https://user-images.githubusercontent.com/1259027/51345381-d856d100-1a68-11e9-95b8-16f354f154e6.png)

## Types of changes
New feature that only affects the settings-sidebar component.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

## Basic Usage Example:
```
const { registerPlugin } = wp.plugins;
const { PluginDocumentSetting } = wp.editPost;

const MyDocumentSettingTest = () => (
	<PluginDocumentSetting className="my-document-setting-plugin">
		<p>My Document Setting Panel</p>
	</PluginDocumentSetting>
);

registerPlugin( 'document-setting-test', { render: MyDocumentSettingTest } );
```

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' of github.com:ryanwelcher/gutenberg

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Create new file for the plugin

Commit:
Adding new slotfill

Commit:
Adds tests

adamsilverstein reviewed (commented):
LGTM! This is an elegant solution to a useful extension point.

Commit:
Ensuring that the slot wraps in a PanelBody and updates tests

Commit:
Adds unintentionally deleted snapshot

youknowriad reviewed (changes requested):
Referring to the related slack discussion. https://wordpress.slack.com/archives/C02QB2JS7/p1545318632348900

Let's wait a bit before introducing this as Phase 2 work could have an impact on backward compatibilty here.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

gziolo commented:
@youknowriad any new thoughts on this proposal? Code wise it looks solid. There is only the question related to the UI considerations and how does it fit in the wider picture of the block editor on other screens like the widget one.

Commit:
Merge remote-tracking branch 'upstream/master'

youknowriad commented:
I think we can explore this a bit further now. The original reasons for delaying this are still valid (could become an inspector control in phase 2 if we ever introduce a "Document" or "Post" block) but it feels a bit too seen and this use-case is legit.

Can we expaned the description of the PR with a plugin example to test it?

I think we should rename this to `DocumentSetting` as it's a more semantic name that sets the expectation that this is about adding new document settings. They are now displayed in the document panel but the framework (editor) can display them in a different way if needed.

Also, is there a way to display the plugin icon in the panel header (to clarify the origin of the panel).

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

ryanwelcher commented:
@youknowriad thanks for the follow up on this. Did you want me to include the testing code as part of the branch or just add it to the description here? I'll have the PR updated shortly to update against latest master and change the slot name.

youknowriad commented:
> Did you want me to include the testing code as part of the branch or just add it to the description here?

I was suggesting just add it to the description but if you could add an e2e test with the example, that would be even better :P

ryanwelcher commented:
> I was suggesting just add it to the description but if you could add an e2e test with the example, that would be even better :P

@youknowriad  There is a test with the original PR, do we need others as well?

youknowriad commented:
I was thinking more an e2e test to avoid regressions. I think in general when it comes to UI, e2e tests should be preferred over unit tests.

Commit:
Changes as per Riad

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Exporting the new slot on the EditPost global

Commit:
Adds e2e test for the new slot

Commit:
Fixes PHPCS error with whitespace

ryanwelcher commented:
@youknowriad I've made the changes indicated and added an e2e test. Looks like the build is failing but it seems unrelated to the change here. 

Is there anything else I can do to help move this forward?

gziolo commented on packages/e2e-tests/plugins/plugins-api/document-setting.js:
```diff
@@ -0,0 +1,20 @@
+	var el = wp.element.createElement;
+	var __ = wp.i18n.__;
+	var registerPlugin = wp.plugins.registerPlugin;
+	var PluginDocumentSetting = wp.editPost.PluginPostStatusInfo;
```
This is incorrect. It pulls a different fill.

gziolo commented on packages/e2e-tests/specs/plugins/plugins-api.test.js:
```diff
@@ -76,4 +76,12 @@ describe( 'Using Plugins API', () => {
+	describe( 'Document Setting Custom Panel', () => {
+		it( 'Should render a custom panel inside Document Setting sidebar', async () => {
+			await openDocumentSettingsSidebar();
+			const pluginPostStatusInfoText = await page.$eval( '.edit-post-sidebar .my-document-setting-plugin', ( el ) => el.innerText );
```
`pluginPostStatusInfoText` should be renamed. I don't think this test is solid enough as it didn't catch the issue with the wrong fill component defined.

gziolo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -30,17 +31,22 @@ const SettingsSidebar = ( { sidebarName } ) => (
-					<PageAttributes />
-					<MetaBoxes location="side" />
-				</Fragment>
+				<PluginDocumentSetting.Slot>
```
What was the reason for using render function which wraps all existing panels? It can be also defined in place of `{ fills }` as `<PluginDocumentSetting.Slot />`.

ryanwelcher replied:
This was modeled after the approach taken for the `<PostStatus />` component. https://github.com/WordPress/gutenberg/blob/master/packages/edit-post/src/components/sidebar/post-status/index.js#L29

If this is incorrect, I'm happy to change it but I think we should standardize whatever approach is taken.

Out of curiosity, can you demonstrate what that looks like and explain any differences/benefits to the other approach?

gziolo commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -30,17 +31,22 @@ const SettingsSidebar = ( { sidebarName } ) => (
+					{ ( fills ) => (
+						<Fragment>
+							<PostStatus />
+							{ fills }
```
Out of curiosity, why do fills are placed after `<PostStatus />`? Why not before `<MetaBoxes location="side" />` or any other place?

ryanwelcher replied:
The placement was chosen based on a pretty common request ( at least where I work ) and it seemed appropriate to place it below PostStatus as most items being added would be of less importance than the data displayed there.

In a perfect world, being able to insert items between all items would be ideal but this is a good start.

gziolo commented on packages/edit-post/src/components/sidebar/plugin-document-setting/index.js:
```diff
@@ -0,0 +1,71 @@
+ * 		},
+ * 		__( 'My Document Setting Panel' )
+ * 		);
+ * 	}
```
Nit: missing semicolon:
```suggestion
 * 	};
```

gziolo commented on packages/edit-post/README.md:
```diff
@@ -100,6 +100,56 @@ _Returns_
+const { registerPlugin } = wp.plugins;
+const { PluginDocumentSetting } = wp.editPost;
+
+const MyDocumentSettingTest = () => (
```
Nit: Whitespaces need to be improved in this code snippet.

gziolo commented on packages/edit-post/README.md:
```diff
@@ -100,6 +100,56 @@ _Returns_
+		},
+		__( 'My Document Setting Panel' )
+		);
+	}
```
Nit: Whitespaces need to be improved in this code snippet.

gziolo commented on packages/edit-post/src/components/sidebar/plugin-document-setting/index.js:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @param {Object} props Component properties.
+ * @param {string} [props.className] An optional class name added to the row.
+ *
```
There are 2 more props to document: `title` and `initialOpen`.

youknowriad commented:
> Also, is there a way to display the plugin icon in the panel header (to clarify the origin of the panel).

Can we implement this?

Commit:
Updates to whitespace and docs as per gziolo

Commit:
Docs updates

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Uses correct slotfill.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Simplifies the Slot render

Commit:
Updates Slot show the icon if passed as part of register_plugin. Ensures that each fill has a PanelBody

Commit:
Updates e2e tests

ryanwelcher commented:
> > Also, is there a way to display the plugin icon in the panel header (to clarify the origin of the panel).
> 
> Can we implement this?

@youknowriad I have added this in.
![panel_with_icon](https://user-images.githubusercontent.com/1259027/58898311-50e0c600-86c8-11e9-9127-69dc5f9e542c.png)

@gziolo I have also updated the tests and simplified the Slot render.

ryanwelcher commented:
This new slot will render a new panel per item registered. In order to get the toggle to save state, the plugin must register the callbacks needed. If none are registered, then the panel will still open/close but the state will not persist.

Here is a snippet I am using to test locally:
```const { registerPlugin } = wp.plugins;
const { PluginDocumentSetting } = wp.editPost;
const { compose } = wp.compose;
const { withDispatch, withSelect} = wp.data;

const MyDocumentSettingTest = ( { isOpened, onTogglePanel } ) => (
	<PluginDocumentSetting
		title={"My Title"}
		className={"my-awesome-class"}
		icon={'smiley'}
		isOpened={ isOpened }
		onTogglePanel={onTogglePanel}
	>
		My Document Setting Panel
	</PluginDocumentSetting>
);
const FIRST_PANEL = 'ryan-test-one';
const MyDocumentPanel = compose(
	withSelect( ( select ) => {
		return (
			{ isOpened: select( 'core/edit-post' ).isEditorPanelOpened( FIRST_PANEL ) }
		);
	} ),
	withDispatch( ( dispatch ) => ( {
		onTogglePanel() {
			return dispatch( 'core/edit-post' ).toggleEditorPanelOpened( FIRST_PANEL );
		},
	} ) ),
)( MyDocumentSettingTest );
registerPlugin( 'document-setting-test', { render: MyDocumentPanel });


const MyDocumentSettingTest2 = ( { isOpened, onTogglePanel } ) => (
	<PluginDocumentSetting
		title={"My Title 2"}
		className={"my-awesome-class-2"}
		isOpened={ isOpened }
		onTogglePanel={ onTogglePanel }
	>
		My Document Setting Panel 2
	</PluginDocumentSetting>
);
const SECOND_PANEL = 'ryan-test-two';
const MyDocumentPanel2 = compose(
	withSelect( ( select ) => {
		return (
			{ isOpened: select( 'core/edit-post' ).isEditorPanelOpened( SECOND_PANEL ) }
		);
	} ),
	withDispatch( ( dispatch ) => ( {
		onTogglePanel() {
			return dispatch( 'core/edit-post' ).toggleEditorPanelOpened( SECOND_PANEL );
		},
	} ) ),
)( MyDocumentSettingTest2 );

registerPlugin( 'document-setting-test-2', { render:MyDocumentPanel2, icon: 'twitter'  });

const MyDocumentSettingTest3 = () => (
	<PluginDocumentSetting
		title={"My Title 3"}
		className={"my-awesome-class-2"}
	>
		My Document Setting Panel 3
	</PluginDocumentSetting>
);
registerPlugin( 'document-setting-test-3', { render: MyDocumentSettingTest3, icon: 'twitter'  });`

gziolo reviewed (commented):
> This new slot will render a new panel per item registered. In order to get the toggle to save state, the plugin must register the callbacks needed. If none are registered, then the panel will still open/close but the state will not persist.

I didn't think of it at all initially. Ideally, this should be included in the implementation so it works the same way for all panels. In addition, it should be also wrapped with the logic which allows removing such panels programatically using `wp.data.dispatch( 'core/edit-post' ).removeEditorPanel( 'my-panel-name' );`

ryanwelcher commented:
> > This new slot will render a new panel per item registered. In order to get the toggle to save state, the plugin must register the callbacks needed. If none are registered, then the panel will still open/close but the state will not persist.
> 
> I didn't think of it at all initially. Ideally, this should be included in the implementation so it works the same way for all panels. In addition, it should be also wrapped with the logic which allows removing such panels programatically using `wp.data.dispatch( 'core/edit-post' ).removeEditorPanel( 'my-panel-name' );`

@gziolo I agree - I was not able to get that working, however. I couldn't get a reference to the FIll when I used `compose` on the Slot to be able to get a unique panel identifier. I'll try again but any help would be appreciated.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Updates src order

Commit:
Updates docs

Commit:
Updated docs

ryanwelcher commented on packages/edit-post/src/index.js:
```diff
@@ -103,3 +103,4 @@ export { default as PluginPostStatusInfo } from './components/sidebar/plugin-pos
 export { default as PluginPrePublishPanel } from './components/sidebar/plugin-pre-publish-panel';
 export { default as PluginSidebar } from './components/sidebar/plugin-sidebar';
 export { default as PluginSidebarMoreMenuItem } from './components/header/plugin-sidebar-more-menu-item';
+export { default as PluginDocumentSetting } from './components/sidebar/plugin-document-setting';
```
I'm not opposed to the name change - the more descriptive ( within reason) the better I think.

gziolo replied:
We would have to change the name of the file as well. Let's do it for consistency given that we both agree that we should also apply the default behavior of Document Settings panel - toggle state stored in the local storage for subsequent visits.

ryanwelcher commented on packages/edit-post/README.md:
```diff
@@ -100,6 +100,10 @@ _Returns_
 
 -   `WPElement`: The WPElement to be rendered.
 
+<a name="PluginDocumentSetting" href="#PluginDocumentSetting">#</a> **PluginDocumentSetting**
```
I've moved this around and made some updates.

Commit:
Removes unneeded mock and component

ryanwelcher commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -29,8 +31,9 @@ const SettingsSidebar = ( { sidebarName } ) => (
 		<Panel>
 			{ sidebarName === 'edit-post/document' && (
-				<>
+				<Fragment>
```
Removed. This was left over from some testing - sorry!

ryanwelcher commented on packages/edit-post/src/components/sidebar/plugin-document-setting/test/index.js:
```diff
@@ -0,0 +1,32 @@
+ */
+import PluginDocumentSetting from '../';
+
+jest.mock( '../../../../../../components/src/button' );
```
Removed

gziolo commented:
> I agree - I was not able to get that working, however. I couldn't get a reference to the FIll when I used `compose` on the Slot to be able to get a unique panel identifier. I'll try again but any help would be appreciated.

I will try to investigate what would be necessary to get it done on Monday and let you know. I know that `PluginSidebar` takes `name` prop to provide the unique name after combining it with the name of the registered plugin:
https://github.com/WordPress/gutenberg/blob/2ab29100b413feb4932cbd576c9e7a9bde31c18e/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js#L81
https://github.com/WordPress/gutenberg/blob/2ab29100b413feb4932cbd576c9e7a9bde31c18e/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js#L135-L139

I bet, we will need something similar here as well.

gziolo commented on packages/edit-post/src/components/sidebar/plugin-document-setting/index.js:
```diff
@@ -0,0 +1,86 @@
+		<PanelBody
+			className={ className }
+			title={ title }
+			opened={ isOpened }
```
If we decide to replicate the logic with toggling used for other panels, this props shouldn't be exposed but rather leave it as an internal thing. We might offer `initialOpen` (https://github.com/WordPress/gutenberg/blob/4371354d77c8b39c99628c590effd560dbe795a3/packages/components/src/panel/README.md#initialopen) instead but it might be confusing as it would make sense to apply it only until it gets toggled and persisted in the browser's storage. There still might be value to it 🤷‍♂

ryanwelcher replied:
I removed it for now and left it internal.

Commit:
Fixes misnamed export and adds toggle handling to the slot

Commit:
Cleaning up the Slot properties

Commit:
Doc updates

Commit:
Doc updates 2: Docs harder

Commit:
Adds logic to allow removal of panel using removeEditorPanel

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Removes snapshot tests due to strange failures

Commit:
Updates e2e test

ryanwelcher commented:
@gziolo I was able to get the toggle controls working - thanks for the input! I also believe I have addressed all of your notes.

gziolo reviewed (commented):
This looks solid and is nearly ready to go. I left some comments with nitpicks so we ensure it's properly documented.

There is also a separate issue which we need to consider in the follow-up - how to expose those registered panels in the Options modal:

![Screen Shot 2019-06-17 at 12 02 08](https://user-images.githubusercontent.com/699132/59596336-cc466c80-90f7-11e9-9cd4-818a906ffd1c.png)

This is challenging on its own, but we can always refactor `PluginDocumentSettingFill` to render two fills, where one targets the sidebar and another one would render to the Options modal :) Prior art for the pinned plugins feature:
https://github.com/WordPress/gutenberg/blob/1d653d1694164f51c5bef67afc3bfdb25870eddd/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js#L33-L43

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

ryanwelcher commented on packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js:
```diff
@@ -0,0 +1,104 @@
+	return (
+		<Fill>
+			<PanelBody
+				name={ name }
```
@gziolo All good points, I removed the `{...props}` and only allow the props that are documented.

ryanwelcher commented on packages/edit-post/src/components/sidebar/plugin-document-setting-panel/index.js:
```diff
@@ -0,0 +1,104 @@
+ * @param {Object} props Component properties.
+ * @param {string} [props.name] The machine-friendly name for the panel.
+ * @param {string} [props.className] An optional class name added to the row.
+ * @param {string} [props.title] The title of the panel
```
Updated

Commit:
Explicitly using props instead of ...props

Commit:
Docs updated and correctly naming import

ryanwelcher commented on packages/edit-post/src/components/sidebar/settings-sidebar/index.js:
```diff
@@ -20,6 +20,7 @@ import PostLink from '../post-link';
 import DiscussionPanel from '../discussion-panel';
 import PageAttributes from '../page-attributes';
 import MetaBoxes from '../../meta-boxes';
+import PluginDocumentSetting from '../plugin-document-setting-panel';
```
Nice catch, I've updated.

Commit:
Updates tests

ryanwelcher commented on packages/e2e-tests/specs/plugins/plugins-api.test.js:
```diff
@@ -76,4 +76,12 @@ describe( 'Using Plugins API', () => {
+		it( 'Should render a custom panel inside Document Setting sidebar', async () => {
+			await openDocumentSettingsSidebar();
+			const pluginDocumentSettingsText = await page.$eval( '.edit-post-sidebar .my-document-setting-plugin', ( el ) => el.innerText );
+			expect( pluginDocumentSettingsText ).toBe( 'My Document Setting Panel' );
```
I've made an update here but am a little unclear still - can you confirm this is what you were looking for @gziolo ?

gziolo replied:
E2e test fails in the current shape, I think it's fine to remove line 84 as this should be part of the snapshot as well. I thought about something like:

```js
const pluginSidebarContent = await page.$eval( '.edit-post-sidebar .my-document-setting-plugin', ( el ) => el.innerHTML );
expect( pluginSidebarContent ).toMatchSnapshot();
```
```

gziolo commented:
E2e test fails with the latest changes:

https://travis-ci.com/WordPress/gutenberg/jobs/209846043#L879

```
FAIL  packages/e2e-tests/specs/plugins/plugins-api.test.js (19.869s)
  ● Using Plugins API › Document Setting Custom Panel › Should render a custom panel inside Document Setting sidebar
    expect(received).toBe(expected) // Object.is equality
    Expected: "My Document Setting Panel"
    Received: "
    My Custom Panel
```

gziolo commented:
@youknowriad, I think this PR addresses your comments, can you do sanity check? There is still e2e test which needs to be fixed which should be straightforward. We might also want to include those custom panels in the Options dialog so users could disable them the same way they can do it for all other panels.

youknowriad commented:
> We might also want to include those custom panels in the Options dialog so users could disable them the same way they can do it for all other panels.

Yeah, this seems like something that is needed. I'm fine if it's a separate PR though. Whatever works best for you all.

youknowriad commented:
Tested this, it works great. A few questions:

 - Will we need an id to "enable/disable" the panel or is the plugin's identifier sufficient?
 - It does feel like we may need a way to define the position (or at least a priority or something like that) for the panel. (Also something that we can explore separately).

ryanwelcher commented:
@youknowriad I was able to remove the custom panels using removePanel and looking up the id in the markup.

I have been thinking about the priority aspect as well, I was thinking that adding it to `registerPlugin()` may make sense. This will give us the flexibility to add it to other SlotFills in the future and also for the existing ones such as `PluginPrePublishPanel` and `PluginPostPublishPanel` - both of which could benefit from it.

chrisvanpatten commented on packages/edit-post/README.md:
```diff
@@ -100,6 +100,60 @@ _Returns_
+
+	registerPlugin( 'my-document-setting-plugin', {
+	render: MyDocumentSettingPlugin
+	} );
```
The spacing in here is a bit awkward. Please don't let that block a merge — I am DYING for this feature — but wanted to point it out in case it's a quick fix :)

ryanwelcher replied:
@chrisvanpatten thanks for mentioning this. I've updated :)

Commit:
Fixes doc indents

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' into feature/settings-panel-slot

Commit:
Updates e2e test to use snapshot

Commit:
Updates snapshot for plugins

ryanwelcher commented:
@gziolo @youknowriad I've updated the e2e test and everything is passing. Is there anything else that needs to be addressed on this?

ryanwelcher commented:
@youknowriad I have created #16384 as an approach to manage priority. Any feedback is apprectiated!

youknowriad commented on packages/edit-post/src/index.js:
```diff
@@ -88,6 +88,7 @@ export function initializeEditor( id, postType, postId, settings, initialEdits )
 }
 
 export { default as PluginBlockSettingsMenuItem } from './components/block-settings-menu/plugin-block-settings-menu-item';
+export { default as PluginDocumentSettingPanel } from './components/sidebar/plugin-document-setting-panel';
```
@aduth any thougs on the naming?

aduth replied:
Seems fine to me as proposed.
